### PR TITLE
Feature/consul acl deny

### PR DESCRIPTION
**Pull request by ChrisAubuchon**
Convert Consul ACL default policy from 'allow' to 'deny'. It sets up three new ACLs in Consul: 

| token | k/v path | use |
| ----- | -------- | --- |
| consul_acl_secure_token | '/v1/kv/secure' | Nginx proxy variables |
| consul_acl_marathon_token | '/v1/kv/marathon' | Marathon-consul |
| consul_acl_vault_token | '/v1/kv/vault' | Vault |

The agent acl has write access to the entire k/v store except for the paths listed above.

### CODE REVIEW

**Comment by siddharthist** [comment]
A cursory test looks good on Vagrant.

**Comment by siddharthist** [comment]
Works on GCE.

File: roles/docker/tasks/main.yml [hunk 1]
```diff
@@ -9,23 +9,11 @@
     - docker
     - bootstrap
 
-- name: install docker repo
```
Comment (ryane) [inline roles/docker/tasks/main.yml]: I don't think we want to revert the changes in #597.

**Comment by ryane** [comment]
don't want to check in `roles/handlers/defaults/.main.yml.swp`

**Comment by stevendborrelli** [comment]
Pushing back to 0.5

File: roles/handlers/defaults/main.yml [hunk 2]
```diff
@@ -0,0 +1,4 @@
+consul_nginx_image: ciscocloud/nginx-consul
+consul_nginx_image_tag: latest
```
Comment (ryane) [inline roles/handlers/defaults/main.yml]: should we pin to a specific version?

File: roles/haproxy/defaults/main.yml [hunk 3]
```diff
@@ -1,6 +1,6 @@
 ---
 haproxy_image: ciscocloud/haproxy-consul
-haproxy_image_tag: 0.2
+haproxy_image_tag: latest
```
Comment (ryane) [inline roles/haproxy/defaults/main.yml]: same here?

**Comment by ryane** [comment]
@ChrisAubuchon @BrianHicks is the current plan to merge this instead of #900? Both have merge conflicts right now but want to confirm this is the one we should focus on.

**Comment by ChrisAubuchon** [comment]
@ryane Either way works for me. Have to fix the merge conflicts either way

**Comment by ryane** [comment]
with that last commit everything seems to be working well in this pr. I think this is ready for testing (fresh install and upgrade) once the merge conflict is resolved.

File: roles/consul/tasks/main.yml [hunk 4]
```diff
@@ -68,36 +57,16 @@
...
-- name: install consul scripts
+- name: install scripts
   sudo: yes
   copy:
-    src: "{{ item.src }}"
-    dest: "{{ item.dest }}"
-    mode: 0755
+    src: "{{ item }}"
+    dest: /usr/local/bin/
+    mode: u=rwx,g=rx,o=rx
```
Comment (siddharthist) [inline roles/consul/tasks/main.yml]: We might want to mirror the notation of #944 to keep things consistent here.

**Comment by siddharthist** [comment]
Other than my comment on the permissions, this LGTM! It can be tested w/o that chnage as well.

**Comment by ryane** [comment]
Yep, the changes in #944 is where the merge conflict is

**Comment by TanyaCouture** [comment]
Successfully installed with:
Vagrant
Openstack

**Comment by ryane** [comment]
@TanyaCouture @sehqlr can we also test this against an existing 0.5.1 cluster?

**Comment by siddharthist** [comment]
I got the following failure on DigitalOcean: 
```
TASK: [consul | create secure acl] ******************************************** 
fatal: [lb-mi-edge-01] => One or more undefined variables: 'consul_acl_secure_token' is undefined

FATAL: all hosts have already failed -- aborting
```

**Comment by siddharthist** [comment]
Works on DigitalOcean. All UIs looking good, confirmed that ACL policy is deny, configurations rendering correctly, etc.

**Comment by siddharthist** [comment]
I've been trying to test an upgrade from v0.5.1, but I keep getting failed consul leader elections when deploying that tag. If someone else could try on another cloud provier, we could merge this.

**Comment by ryane** [comment]
@siddharthist are you running the upgrade-packages.yml before installing 0.5.1? I am able to install 0.5.1 (on aws) but I get failed consul leader elections when upgrading from this branch.

```
failed: [resching-aws-control-03] => {"changed": true, "cmd": ["/usr/local/bin/consul-wait-for-leader.sh"], "delta": "0:00:35.079217", "end": "2016-01-10 03:17:25.366237", "rc": 1, "start": "2016-01-10 03:16:
50.287020", "warnings": []}
stdout: No Consul leader elected in 30 seconds
failed: [resching-aws-edge-01] => {"changed": true, "cmd": ["/usr/local/bin/consul-wait-for-leader.sh"], "delta": "0:00:35.161713", "end": "2016-01-10 03:17:25.461280", "rc": 1, "start": "2016-01-10 03:16:50.
299567", "warnings": []}
stdout: No Consul leader elected in 30 seconds
failed: [resching-aws-control-01] => {"changed": true, "cmd": ["/usr/local/bin/consul-wait-for-leader.sh"], "delta": "0:00:35.079081", "end": "2016-01-10 03:17:25.455685", "rc": 1, "start": "2016-01-10 03:16:
50.376604", "warnings": []}
stdout: No Consul leader elected in 30 seconds
```

**Comment by BrianHicks** [comment]
In 04e8be8 I updated Consul to 0.6.1, which removes the separate UI directory install. Should simplify things for us. I tested to make sure it came up OK on GCE, Consul looks fine to me.

**Comment by ChrisAubuchon** [comment]
Broken on Vagrant:
```
Jan 11 16:17:58 default consul[4682]: ==> Error starting agent: Failed to configure keyring: mkdir /var/lib/consul/serf: permission denied
```
I have a PR for this: https://github.com/asteris-llc/mantl-packaging/pull/16/files

**Comment by ryane** [comment]
@ChrisAubuchon it appears running this against an existing cluster results in quorum loss. Any idea what we can do to orchestrate a successful upgrade?

(note: test prior to 04e8be8)

**Comment by ChrisAubuchon** [comment]
@ryane serial=1 maybe? Is ansible upgrading and restarting them all at the same time?

**Comment by ryane** [comment]
with the latest commits, the consul UI made available by nginx-consul (https://control-node:8500/ui/) returns 

```
HTTP error code from Consul: 403 Forbidden
```

Is this expected? The UI proxied by nginx-mantlui is working correctly.

**Comment by ChrisAubuchon** [comment]
Looks like a Consul bug. The UI collects data using the `/v1/internal/ui/` endpoint. Using the external UI code, the request looks like: 
```
"GET /consul/v1/internal/ui/services?dc=vagrant&token= HTTP/1.1"
```
The internal UI is failing with:
```
[ERR] http: Request GET /v1/internal/ui/nodes?dc=vagrant&token=<hidden>, error: ACL not found from=127.0.0.1:46054
```
After some digging, the internal UI is sending the request as:
```
"GET /v1/internal/ui/nodes?dc=vagrant&token=undefined HTTP/1.1"
```
which fails because there is no ACL `undefined`. Creating an ACL token with
```shell
$ consul-cli acl-update --token=<token> --rule='key::deny' --rule='service::deny' undefined
```
and the Consul UI works again.

**Comment by siddharthist** [comment]
We should raise an issue upstream and document this behavior before merging.

**Comment by ChrisAubuchon** [comment]
My dream of having a semi-relevant PR open for a full year lives on!

**Comment by ryane** [comment]
hooray! looks like a similar issue to https://github.com/hashicorp/consul/issues/1316

**Comment by ChrisAubuchon** [comment]
At this point I think we close this PR. The gain is minimal and the merge process has been a nightmare.

**Comment by BrianHicks** [comment]
Closing it is fine with me, but then we need to cherry-pick my commits back here to #900. Just noting so we don't forget that. ;)

**Comment by ChrisAubuchon** [comment]
I've added the consul-ui package back in to work around the broken internal UI.

**Comment by stevendborrelli** [comment]
What are the blockers for getting this merged?

**Comment by ChrisAubuchon** [comment]
For starters it has to be rebased (again).

**Comment by ryane** [comment]
lgtm. deployed on GCE and am not seeing any adverse effects. the /secure tree in the KV store is no longer visible in the UI (i.e. acls are working). Note: you must re-generate security.yml to deploy this successfully.
